synapse/handlers/read_marker.py

```diff
@@ -46,12 +47,21 @@ async def received_client_read_marker(
             )
 
             should_update = True
+            # Get event ordering, this also ensures we know about the event
+            event_ordering = await self.store.get_event_ordering(event_id)
 
             if existing_read_marker:
-                # Only update if the new marker is ahead in the stream
-                should_update = await self.store.is_event_after(
-                    event_id, existing_read_marker["event_id"]
-                )
+                try:
+                    old_event_ordering = await self.store.get_event_ordering(
+                        existing_read_marker["event_id"]
+                    )
+                except SynapseError:
+                    # Old event no longer exists, assume new is ahead. This may
+                    # happen if the old event was removed due to retention.
+                    pass
```

Comment:
Would it be possible to add a test for this?

Reply:
Yep! https://github.com/matrix-org/synapse/pull/15464/commits/cbfc04efaf2df0f41dc1e93c4dd112c9ebb9f54b